File: internal/repository/supplier.go
package repository

import (
	"time"
	"context"

	"github.com/uptrace/bun"
	"warehouse/internal/model"
)

type SupplierRepository struct {
	db *bun.DB
}

func NewSupplierRepository(db *bun.DB) *SupplierRepository {
	return &SupplierRepository{db: db}
}

type SupplierQueryFilter struct {
	Code     string
	Name     string
	Contact  string
	Phone    string
	Status   *int
	Page     int
	PageSize int
}

func (r *SupplierRepository) Create(ctx context.Context, supplier *model.Supplier) error {
	_, err := r.db.NewInsert().Model(supplier).Exec(ctx)
	return err
}

func (r *SupplierRepository) GetByID(ctx context.Context, id int64) (*model.Supplier, error) {
	supplier := new(model.Supplier)
	err := r.db.NewSelect().
		Model(supplier).
		Where("id = ?", id).
		Where("deleted_at IS NULL").
		Scan(ctx)
	if err != nil {
		return nil, err
	}
	return supplier, nil
}

func (r *SupplierRepository) GetByCode(ctx context.Context, code string) (*model.Supplier, error) {
	supplier := new(model.Supplier)
	err := r.db.NewSelect().
		Model(supplier).
		Where("code = ?", code).
		Where("deleted_at IS NULL").
		Scan(ctx)
	if err != nil {
		return nil, err
	}
	return supplier, nil
}

func (r *SupplierRepository) List(ctx context.Context, filter *SupplierQueryFilter) ([]model.Supplier, int, error) {
	var suppliers []model.Supplier

	q := r.db.NewSelect().
		Model(&suppliers).
		Where("deleted_at IS NULL")

	if filter.Code != "" {
		q = q.Where("code LIKE ?", "%"+filter.Code+"%")
	}
	if filter.Name != "" {
		q = q.Where("name LIKE ?", "%"+filter.Name+"%")
	}
	if filter.Contact != "" {
		q = q.Where("contact LIKE ?", "%"+filter.Contact+"%")
	}
	if filter.Phone != "" {
		q = q.Where("phone LIKE ?", "%"+filter.Phone+"%")
	}
	if filter.Status != nil {
		q = q.Where("status = ?", *filter.Status)
	}

	total, err := q.
		Order("id DESC").
		Offset((filter.Page - 1) * filter.PageSize).
		Limit(filter.PageSize).
		ScanAndCount(ctx)
	if err != nil {
		return nil, 0, err
	}
	return suppliers, total, nil
}

func (r *SupplierRepository) Update(ctx context.Context, supplier *model.Supplier) error {
	_, err := r.db.NewUpdate().
		Model(supplier).
		WherePK().
		Where("deleted_at IS NULL").
		Exec(ctx)
	return err
}

func (r *SupplierRepository) Delete(ctx context.Context, id int64) error {
	_, err := r.db.NewUpdate().
		Model((*model.Supplier)(nil)).
		Set("deleted_at = ?", time.Now()).
		Where("id = ?", id).
		Where("deleted_at IS NULL").
		Exec(ctx)
	return err
}

File: web/src/api/supplier.ts
import api from './client'
import type { Supplier, CreateSupplierRequest, UpdateSupplierRequest } from '../types/partner'
import type { PaginatedResponse } from '../types/warehouse'

export interface SupplierFilter {
  code?: string
  name?: string
  contact?: string
  phone?: string
  status?: number
}

export const supplierApi = {
  list: async (page = 1, size = 10, filter?: SupplierFilter): Promise<PaginatedResponse<Supplier>> => {
    const params = new URLSearchParams()
    params.append('page', String(page))
    params.append('size', String(size))
    
    if (filter?.code) {
      params.append('code', filter.code)
    }
    if (filter?.name) {
      params.append('name', filter.name)
    }
    if (filter?.contact) {
      params.append('contact', filter.contact)
    }
    if (filter?.phone) {
      params.append('phone', filter.phone)
    }
    if (filter?.status !== undefined) {
      params.append('status', String(filter.status))
    }
    
    const response = await api.get<PaginatedResponse<Supplier>>('/suppliers', {
      params: params
    })
    return response.data
  },

  get: async (id: number): Promise<Supplier> => {
    const response = await api.get<Supplier>(`/suppliers/${id}`)
    return response.data
  },

  create: async (data: CreateSupplierRequest): Promise<Supplier> => {
    const response = await api.post<Supplier>('/suppliers', data)
    return response.data
  },

  update: async (id: number, data: UpdateSupplierRequest): Promise<Supplier> => {
    const response = await api.put<Supplier>(`/suppliers/${id}`, data)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/suppliers/${id}`)
  }
}
